packages/utils/telemetry-utils/src/sampledTelemetryHelper.ts

```diff
@@ -0,0 +1,131 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IDisposable, ITelemetryLogger, ITelemetryPerformanceEvent } from "@fluidframework/common-definitions";
+import { performance } from "@fluidframework/common-utils";
+
+interface Measurements {
+    latestDuration: number;
+    executionCount: number;
+    totalDuration: number;
+    minDuration: number | undefined;
+    maxDuration: number | undefined;
+}
+
+/**
+ * Helper class that executes a specified code block and writes an
+ * {@link @fluidframework/common-definitions#ITelemetryPerformanceEvent} to a specified logger every time a specified
+ * number of executions is reached (or when the class is disposed). The `duration` field in the telemetry event is
+ * the duration of the latest execution (sample) of the specified function. See the documentation of the
+ * `includeAggregateMetrics` parameter for additional details that can be included.
+ */
+ export class SampledTelemetryHelper implements IDisposable {
+    disposed: boolean = false;
+
+    private readonly measurementsMap = new Map<string, Measurements>();
+
+    /**
+     * @param eventBase - Custom properties to include in the telemetry performance event when it is written.
+     * @param logger - The logger to use to write the telemetry performance event.
+     * @param includeAggregateMetrics - If set to `true`, the telemetry performance event will include aggregated
+     *                                  metrics (execution count, total duration, min duration, max duration) for
+     *                                  all the executions in between generated events.
+     */
+    public constructor(
+        private readonly eventBase: any,
+        private readonly logger: ITelemetryLogger,
+        private readonly includeAggregateMetrics: boolean = false) {
+    }
+
+    /**
+     * @param codeToMeasure - The code to be executed and measured.
+     * @param countThreshold - The telemetry performance event with aggregated data will be
+     *                         generated after this many executions of the code block.
+     * @param dimension - A key to track executions of the code block separately. Each different
+     *                    value of this parameter has a separate set of executions and metrics tracked
+     *                    by the class. If no such distinction needs to be made, do not provide a value.
+     * @returns Whatever the passed-in code block returns.
+     */
+    public measure<T>(
+        codeToMeasure: () => T,
+        countThreshold: number,
+        dimension: string = ""): T {
```

Comment:
there will actually a new column, and for each dimension with only the value that makes the name of the column. i'm not even sure what the default empty string case will do

Reply:
if we want to keep if simple, the constructor could take the name of the dimension, and the measure could take the value. still only one, but that looks like enough for now

Reply:
Yeah, I want to make it into a property bag, just haven't decided how to take that and make it into a key for the map in the helper class. If we have a utility to hash an arbitrary property bag consistently, that could work. Otherwise, iterating over the sorted `Object.keys()` and concatting the values is probably the way to go?

Reply:
> changing the api to remove dimension, and make the class handle a collection, where you can specific sets of named properties that map to a specific measure. similar to above, but forces the consumer to use names, rather than generating them

I was leaning towards this to begin with. It makes life slightly more difficult for consumer classes because they have to maintain the collection of telemetry helpers and branch in their code to use the correct one.

What about having the consumer provide an optional lambda in the helper's constructor that maps the property bags it might provide to keys for the helper's internal map? Still leaves some responsibility to the consumer, to provide a lambda that handles all necessary cases, ideally doesn't fail, and isn't too expensive; but the consumer has more knowledge of what their property bags would look like and if there's a trivial way to make them into unique keys.

Reply:
i think it makes it a lot harder for anyone to pick up and use. generally, people will just look at other usages and copy what they have done. i also worry people could do the wrong things by not understanding how it works.

what if we just take a map of string to ITelemertyProps, so it is one the consumer to build the mapping, and our code just does lookup?